src/ui/StatusEditor.svelte

```diff
@@ -26,9 +26,18 @@
         const taskWithEditedStatusApplied = task.handleNewStatus(selectedStatus).pop();
 
         if (taskWithEditedStatusApplied) {
-            // Update the doneDate field, in case changing the status changed the value:
-            editableTask.doneDate = taskWithEditedStatusApplied.done.formatAsDate();
-            editableTask.cancelledDate = taskWithEditedStatusApplied.cancelled.formatAsDate();
+            // change the done date using XNOR logic:
+            //  done date is empty and new status is DONE
+            // OR
+            //  done date is filled and new status is not DONE
```

Comment:
I have read this comment and the code several times, and have not been able to match the code to the behaviour.

I feel like maybe the code is being clever, and clever code is rarely easy to maintain.

Reply:
Well, I had the same thought... Does this look better?

```ts
           if (
                (editableTask.doneDate === '' && newStatus.isCompleted()) ||
                (editableTask.doneDate !== '' && !newStatus.isCompleted())
            ) {
                editableTask.doneDate = taskWithEditedStatusApplied.done.formatAsDate();
            }

            if (
                (editableTask.cancelledDate === '' && newStatus.isCancelled()) ||
                (editableTask.cancelledDate !== '' && !newStatus.isCancelled())
            ) {
                editableTask.cancelledDate = taskWithEditedStatusApplied.cancelled.formatAsDate();
            }
```

Reply:
> Well, I had the same thought... Does this look better?

A little bit.

- It's confusing that it tests the string before the boolean - I feel like it would be more readable if it tested the boolean first
- It's effort to read and review  because of the repetition. 

I would try extracting variables something like:

```ts
const isInStatus = newStatus.isCompleted();
const hasDate = editableTask.doneDate;
```

Then extract a local function, and you can split the conditions and use early-return to add clear comments about each of the two && options...

Reply:
Interesting, I haven't seen it like that, but I gave it a go and found myself in a pretty neat implementation. I couldn't use early returns though, in my experiments the logic was easier to think about in a reverse situation, please check the jsdoc of `setStatusRelatedDate()`

Reply:
Thank you. I've replied in code review...

Reply:
I experimented with this, removing the need for the JSdoc block...

I am not sure if it is worth keeping, but I find it easier to read...

```ts
    function setStatusRelatedDate(currentValue: string, isInStatus: boolean, editedValue: TasksDate) {
        const dateFieldIsEmpty = currentValue === '';

        if (dateFieldIsEmpty && isInStatus) {
            // the date field is empty and the status was set (set the date from the task with the applied status)
            return editedValue.formatAsDate();
        }

        if (!dateFieldIsEmpty && !isInStatus) {
            // the date field is not empty but another status was set (clean the date field)
            return '';
        }

        return currentValue;
    }
```

```ts
        if (taskWithEditedStatusApplied) {
            editableTask['doneDate'] = setStatusRelatedDate(
                editableTask['doneDate'],
                selectedStatus.isCompleted(),
                taskWithEditedStatusApplied['done'],
            );

            editableTask['cancelledDate'] = setStatusRelatedDate(
                editableTask['cancelledDate'],
                selectedStatus.isCancelled(),
                taskWithEditedStatusApplied['cancelled'],
            );
        }
```

Reply:
What do you think about this order:

```ts
        if (isInStatus && dateFieldIsEmpty) {
            // the date field is empty and the status was set (set the date from the task with the applied status)
            return editedValue.formatAsDate();
        }

        if (!isInStatus && !dateFieldIsEmpty) {
            // the date field is not empty but another status was set (clean the date field)
            return '';
        }
```

Reply:
> I experimented with this, removing the need for the JSdoc block...
> 
> I am not sure if it is worth keeping, but I find it easier to read...

This looks so much better, so I'm happy to retain that =)

Reply:
> What do you think about this order of the `if` conditions?
> 
> ```ts
>         if (isInStatus && dateFieldIsEmpty) {
>             // the date field is empty and the status was set (set the date from the task with the applied status)
>             return editedValue.formatAsDate();
>         }
> 
>         if (!isInStatus && !dateFieldIsEmpty) {
>             // the date field is not empty but another status was set (clean the date field)
>             return '';
>         }
> ```

The order doesn't matter - both cases are covered by separate test cases

Reply:
Ah, I was unclear. I find that testing `isInStatus` first, then checking `dateFieldIsEmpty` feels more natural, given the purpose of the code.

Reply:
(I flipped the && order)

Reply:
Oh, yes, this looks better indeed

Reply:
> Oh, yes, this looks better indeed

Thanks.... Do you want to make that change, or do you want me to make it later? (Laptop battery about to die...)

Reply:
Done